instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/BaseTracer.java

```diff
@@ -30,6 +30,11 @@
   public static final ContextKey<Span> CONTEXT_CLIENT_SPAN_KEY =
       ContextKey.named("opentelemetry-trace-auto-client-span-key");
 
+  // Keeps track of the application root (e.g. servlet context path) that needs to be prepended
```

Comment:
Isn't this too specific a use case to put in BaseTracer?

Reply:
Also is context the only way? I thought servlet has `getServletPath` and `getRequestUri` we'd be able to compute the context path based on it?

https://docs.oracle.com/javaee/6/api/javax/servlet/http/HttpServletRequest.html

Reply:
> Isn't this too specific a use case to put in BaseTracer?

That's a good idea, I can move this to something like an `ApplicationRoot` utility class in `instrumentation-api`

> Also is context the only way? I thought servlet has `getServletPath` and `getRequestUri` we'd be able to compute the context path based on it?

I'm not following this suggestion, can you expand?

Reply:
> I'm not following this suggestion, can you expand?

I also might not be following the problem :) But if the problem is the URL we populate is incomplete, I would assume something on HttpServlerRequest would be sufficient to provide the complete URL?

Reply:
Got it, ya, that's what this PR does. When servlet starts, it captures `request.getContextPath()` and stores in the context, so that later on, when routes are updated, they can pre-pend the context path in front of the route.

Reply:
I see I think I follow - there are places we compute the route and update name, but have no access to the HttpServletRequest. It feels a bit weird to put a string into Context like this, I'd probably put HttpServletRequest itself to indicate we are propagating information about the request. 

Also instead of `ApplicationRoot` utility, I was hoping this could go in HttpServletTracer since it's specific to servlet.

Reply:
> I see I think I follow - there are places we compute the route and update name, but have no access to the HttpServletRequest. It feels a bit weird to put a string into Context like this, I'd probably put HttpServletRequest itself to indicate we are propagating information about the request.

We can't have a `ContextKey<HttpServletRequest>` in `instrumentation-api`, since `instrumentation-api` doesn't have access to `HttpServletRequest`.

We could do `ContextKey<Object>` and cast it when we pull it out, but that doesn't feel great either.

> Also instead of `ApplicationRoot` utility, I was hoping this could go in HttpServletTracer since it's specific to servlet.

It needs to be accessed from Tracers which don't extend HttpServletTracer.

We could still put it in HttpServletTracer, and access it from those Tracers, do u think that's better? I don't have strong opinion.

Reply:
Perhaps we could have a servlet-utils with a static helper then - actually I had no idea webflux works with servlet since our instrumentation doesn't have a dependency on it! So it seems to make sense that we could have a package dedicated to helping out servlet implementations. Having it outside the main instrumentation API would make me a lot less anxious.

Reply:
That being said if my below suggestions make sense I may have been able to avoid the propagation requirement 🤞

Reply:
I think we need propagation at least for JAX-RS.

I agree there's danger of making it look too general, and people thinking that they need to always use it for all instrumentation.

A new module `:instrumentation:servlet:servlet-context-path` that can be imported into other instrumentation as needed seems nice.

Though if the initial filter/servlet isn't in the same class loader as the jax-rs resource, that may not work, because we will inject the class w/ context key into each place, and they won't share.

What do you think of putting the context key in `io.opentelemetry.instrumentation.api.servlet.ContextPath`, to make it look less general?

Or create `servlet-context-path` as a secondary "instrumentation-api" that we include in the bootstrap class loader?

Or...

Reply:
Yeah I'm trying to find a way to avoid extra API for a seemingly very specific use case.

I have a theory that any request handled by servlet always has access to the `HttpServletRequest` somehow :) It's one of the reasons I've been persistent here.

Adding a break point at `updateSpanNames` and see what's available when running the tests has helped find some options I think.

I found for resteasy, we can get `HttpServletRequest` anywhere with this static call `ResteasyProviderFactory.getContextData(HttpServletRequest.class)`

For jersey, it's not `HttpServletRequest` but we seem to have a similar static call `ContextHandler.getCurrentContext().getContextPath()`.

So is it possible for us to use these to solve the problem at the bottom layer? I guess it means introducing e.g., `ResteasyAnnotationsTracer extends JaxRsAnnotationsTracer` for the implementation-specific extraction but doesn't seem too bad.

Reply:
@anuraaga It looks like `ContextHandler.getCurrentContext().getContextPath()` is for jetty not jersey (well, would work for jersey running in jetty).

Reply:
> Is this whole problem specific to servlets? Or for other HTTP Server spans as well? E.g. Ratpack or Ktor or I-dont-know-what-else?

It's specific to any routing framework that runs inside of a servlet container. I think I was trying to generalize it too much to "application root", but better to keep it specific to "servlet context path".